File: VTK_files/VTK_3Darrays.f90
program VTK
    implicit none
    double precision, allocatable, dimension(:):: x,y,z
    double precision,allocatable,dimension(:,:,:)::M3D_1,M3D_2,M3D_3
    double precision :: dx,dy,dz
    integer :: i,j,k,N
    integer :: nx,ny,nz
    N = 5
    nx = N+1; ny = N+1; nz = N+1;
    allocate(M3D_1(N,N,N),M3D_2(N,N,N),M3D_3(N,N,N),x(nx),y(ny),z(nz))
    dx = 1.0d0/(N-1.0d0) 
    dy = dx
    dz = dx
    do i = 1, N+1
        x(i) = (i-1.0d0)*dx
        y(i) = (i-1.0d0)*dy
        z(i) = (i-1.0d0)*dz
    end do
    do k = 1, N
        do j = 1, N
            do i = 1, N
                M3D_1(i,j,k) = i*j*k
                M3D_2(i,j,k) = i*j*k*0.7d0;
                M3D_3(i,j,k) = i*j*k*0.2d0;
            end do
        end do
    end do
    open(unit=1, file ='contour3D.vtk')
    write ( 1,'(a)' )'# vtk DataFile Version 3.0'
    write ( 1,'(a)' ) 'Non-uniform Rectilinear - Rectilinear Grid'
    write ( 1,'(a)' ) 'ASCII'
    write ( 1,'(a)' ) 'DATASET RECTILINEAR_GRID'
    write ( 1,'(a,3i4)' )'DIMENSIONS', nx,ny,nz
    write ( 1,'(a13,1i4,a6)' ) 'X_COORDINATES',nx,' float'
    write ( 1,'(f14.6)' ) (x(i),i=1,nx)
    write ( 1,'(a13,1i4,a6)' ) 'Y_COORDINATES',ny,' float'
    write ( 1,'(f14.6)' ) (y(j),j=1,ny)
    write ( 1,'(a13,1i4,a6)' ) 'Z_COORDINATES',nz,' float'
    write ( 1,'(f14.6)' ) (z(k),k=1,nz)
    write ( 1, '(a, I16)' ) 'CELL_DATA ', (nx-1)*(ny-1)*(nz-1)
    write ( 1, '(a)' ) 'SCALARS array_A float'
    write ( 1, '(a)' ) 'LOOKUP_TABLE default'
    write ( 1, '(f22.14)' ) (((M3D_1(i,j,k),i=1,(nx-1)),j=1,(ny-1)),k=1,(nz-1))
    write ( 1, '(a)' ) 'SCALARS array_B float'
    write ( 1, '(a)' ) 'LOOKUP_TABLE default'
    write ( 1, '(f22.14)' ) (((M3D_2(i,j,k),i=1,(nx-1)),j=1,(ny-1)),k=1,(nz-1))
    write ( 1, '(a)' ) 'SCALARS array_C float'
    write ( 1, '(a)' ) 'LOOKUP_TABLE default'
    write ( 1, '(f22.14)' ) (((M3D_3(i,j,k),i=1,(nx-1)),j=1,(ny-1)),k=1,(nz-1))    
    deallocate(M3D_1,M3D_2,M3D_3)
end program VTK

File: VTK_files/VTK_3Darrays.c
#include <stdio.h>
#include <string.h>
void VTK(int,int,int,double [*],double [*],double [*],
         double [*][*][*],double [*][*][*],double [*][*][*]);
void treeTOone(int,int,int,double [*][*][*],double [*]);
char a_name[10][100];
int main()
{
    int i,j,k;
    int N=5, nx=N+1,ny=N+1,nz=N+1;
    double M3D_1[nx-1][ny-1][nz-1],
           M3D_2[nx-1][ny-1][nz-1],
           M3D_3[nx-1][ny-1][nz-1];
    double dx,dy,dz,x[nx],y[ny],z[nz];
    dx = 1.0/(N-1.0); 
    dy = dx;
    dz = dx;
    for (i=0;i<=N;i++)
    {
        x[i] = i*dx;
        y[i] = i*dy;
        z[i] = i*dz;
    }
    for (k=0;k<nz-1;k++) 
    {
        for (j=0;j<ny-1;j++) 
        {
            for (i=0;i<nx-1;i++)
            {
                M3D_1[i][j][k] = (i+1)*(j+1)*(k+1);
                M3D_2[i][j][k] = (i+1)*(j+1)*(k+1)*0.7;
                M3D_3[i][j][k] = (i+1)*(j+1)*(k+1)*0.2;
            }
        }
    }
    strcpy(a_name[0],"Array3D.vtk"); //File Name
    strcpy(a_name[1],"array_A");     //Array 1 Flag
    strcpy(a_name[2],"array_B");     //Array 2 Flag
    strcpy(a_name[3],"array_C");     //Array 3 Flag
    VTK(nx,ny,nz,x,y,z,M3D_1,M3D_2,M3D_3);
}
void treeTOone(int nx, int ny, int nz,double array3D[nx][ny][nz],double array1D[nx*ny*nz])
{
    int i,j,k,l=0;
    for (k=0;k<nz;k++) 
    {
        for (j=0;j<ny;j++) 
        {
            for (i=0;i<nx;i++)
            {
                array1D[l] = array3D[i][j][k];
                l++;
            }
        }
    }
}
void VTK(int nx, int ny, int nz,
         double x[nx],double y[ny],double z[nz],
         double array3D_1[nx-1][ny-1][nz-1],
         double array3D_2[nx-1][ny-1][nz-1],
         double array3D_3[nx-1][ny-1][nz-1])
{
    int i,l=0,Ntot=(nx-1)*(ny-1)*(nz-1);
    double array_1D[Ntot];
    FILE* f = fopen(a_name[0],"w");
    fprintf(f, "# vtk DataFile Version 3.0");
    fprintf(f, "\nNon-uniform Rectilinear - Rectilinear Grid");
    fprintf(f, "\nASCII");
    fprintf(f, "\nDATASET RECTILINEAR_GRID");
    fprintf(f, "\nDIMENSIONS   %d   %d   %d",nx,ny,nz);
    fprintf(f, "\nX_COORDINATES   %d float",nx);
    for (i=0;i<nx;i++)
    {
        fprintf(f,"\n%14.6lf",x[i]);
    } 
    fprintf(f, "\nY_COORDINATES   %d float",ny);
    for (i=0;i<ny;i++)
    {
        fprintf(f,"\n%14.6lf",y[i]);
    }  
    fprintf(f, "\nZ_COORDINATES   %d float",nz);
    for (i=0;i<nz;i++)
    {
        fprintf(f,"\n%14.6lf",z[i]);
    }
    fprintf(f, "\nCELL_DATA       %d",Ntot);
    treeTOone(nx-1,ny-1,nz-1,array3D_1,array_1D);
    fprintf(f, "\nSCALARS %s float",a_name[1]);
    fprintf(f, "\nLOOKUP_TABLE default");
    for (i=0; i<Ntot;i++)
    {
        fprintf(f,"\n%22.14lf",array_1D[i]);
    }
    treeTOone(nx-1,ny-1,nz-1,array3D_2,array_1D);
    fprintf(f, "\nSCALARS %s float",a_name[2]);
    fprintf(f, "\nLOOKUP_TABLE default");
    for (i=0; i<Ntot;i++)
    {
        fprintf(f,"\n%22.14lf",array_1D[i]);
    }
    treeTOone(nx-1,ny-1,nz-1,array3D_3,array_1D);
    fprintf(f, "\nSCALARS %s float",a_name[3]);
    fprintf(f, "\nLOOKUP_TABLE default");
    for (i=0; i<Ntot;i++)
    {
        fprintf(f,"\n%22.14lf",array_1D[i]);
    }
    fclose(f);
}
